src/sys/select.rs

```diff
@@ -58,15 +61,30 @@ impl FdSet {
     /// ```
     ///
     /// [`select`]: fn.select.html
-    pub fn highest(&mut self) -> Option<RawFd> {
-        for i in (0..FD_SETSIZE).rev() {
-            let i = i as RawFd;
-            if unsafe { libc::FD_ISSET(i, self as *mut _ as *mut libc::fd_set) } {
-                return Some(i)
-            }
-        }
+    pub fn highest(&self) -> Option<RawFd> {
+        self.fds().next_back()
+    }
 
-        None
+    /// Returns an iterator over the file descriptors in the set.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # extern crate nix;
+    /// # use nix::sys::select::FdSet;
+    /// # use std::os::unix::io::RawFd;
+    /// let mut set = FdSet::new();
+    /// set.insert(4);
+    /// set.insert(9);
+    /// let fds: Vec<RawFd> = set.fds().collect();
+    /// assert_eq!(fds, vec![4, 9]);
+    /// ```
+    #[inline]
+    pub fn fds(&self) -> Fds {
```

Comment:
Typically a C programmer will know what the maximum file descriptor he registered was, and when inspecting the `fd_set` he won't need to iterate all the way to `FD_SETSIZE`, which can be up to 64k on some platforms.  But as written your iterator will always iterate that high.  That's not good for performance.  What if you add an `max: Option<RawFd>` argument to`fds`?  If supplied, the iterator won't check any higher than that.  Such an argument could also be useful to `highest`.

Reply:
Added to `fds()`. I didn't add to `highest()` in the interest of not introducing unnecessary breakage. If someone wants to get that performance gain, they can use `set.fds(Some(my_limit)).next_back()`, which isn't terribly hard to write and is pretty clear.